meter/lgpcs/lgpcs.go

```diff
@@ -120,53 +122,99 @@ func (m *Com) Login() error {
 }
 
 // Data gives the data read from the pcs.
-func (m *Com) Data() (EssData8, error) {
+func (m *Com) Data() (EssData, error) {
 	return m.dataG()
 }
 
 // refreshData reads data from lgess pcs. Tries to re-login if "405" auth_key expired is returned
-func (m *Com) refreshData() (EssData8, error) {
+func (m *Com) refreshData() (EssData, error) {
+	var data EssData
+	var err error
+	if m.essType == LgEss8 {
+		var meterResponse MeterResponse8
+		if err = refreshDataFromLgess[MeterResponse8](&meterResponse, m); err != nil {
+			return EssData{}, err
+		}
+		essDataFromMeterResponse8(&data, &meterResponse)
+	} else {
+		var meterResponse MeterResponse15
+		if err = refreshDataFromLgess[MeterResponse15](&meterResponse, m); err != nil {
+			return EssData{}, err
+		}
+		essDataFromMeterResponse15(&data, &meterResponse)
+	}
+
+	return data, nil
+}
+
+// read data from Lgess, reconnects if session expired
+func refreshDataFromLgess[T MeterResponse8 | MeterResponse15](meterResponse *T, m *Com) error {
 	data := map[string]interface{}{
 		"auth_key": m.authKey,
 	}
 
 	req, err := request.New(http.MethodPost, m.uri+MeterURI, request.MarshalJSON(data), request.JSONEncoding)
 	if err != nil {
-		return EssData8{}, err
+		return err
 	}
 
-	var resp MeterResponse8
-
-	if err = m.DoJSON(req, &resp); err != nil {
+	if err := m.DoJSON(req, meterResponse); err != nil {
 		// re-login if request returns 405-error
 		var se request.StatusError
 		if errors.As(err, &se) && se.StatusCode() == http.StatusMethodNotAllowed {
-			err = m.Login()
+			if err = m.Login(); err != nil {
+				return err
+			}
 
-			if err == nil {
-				data["auth_key"] = m.authKey
-				req, err = request.New(http.MethodPost, m.uri+MeterURI, request.MarshalJSON(data), request.JSONEncoding)
+			data["auth_key"] = m.authKey
+			if req, err = request.New(http.MethodPost, m.uri+MeterURI, request.MarshalJSON(data), request.JSONEncoding); err != nil {
+				return err
 			}
 
-			if err == nil {
-				err = m.DoJSON(req, &resp)
+			if err = m.DoJSON(req, meterResponse); err != nil {
+				return err
 			}
 		}
 	}
 
-	if err != nil {
-		return EssData8{}, err
-	}
+	return err
+}
 
-	res := resp.Statistics
-	if resp.Direction.IsGridSelling > 0 {
-		res.GridPower = -res.GridPower
+// convert response from LgEss8/10 into interface EssData
+func essDataFromMeterResponse8(essData *EssData, meterResponse *MeterResponse8) {
+	essData.GridPower = meterResponse.Statistics.GridPower
```

Comment:
This feels a bit awkward. An alternative would be to make  EssData8 and EssData15 implement a common interface?